### PR TITLE
Add profiling to starcoder

### PR DESCRIPTION
**Pull request by pawalt**
This commit adds profiling to starcoder.

### CODE REVIEW

**Comment by pawalt** [comment]
I added this profiling so I could find out why my `SftTrainer` runs were scaling sub-linearly with node count.

2 nodes:
image
4 nodes:
image

The 2 node steps take about 41ms while the 4 node runs  take about 23ms. This means we're only getting a 1.78x speedup where we'd expect a 2x speedup.

Looking at the traces, this is due to the constant cost of data loading. We see consistent gaps of 4s where the CPU and GPU go unused. This is because I'm (foolishly) loading the data in the same thread as I'm doing the training steps. I tried to split these out, but it's surprisingly challenging. I will do this soon.

If you take out the 4s constant term, you get 37ms of working time for the 2 node run and 19ms of working time for the 4 node run. This amounts to a 1.94x speedup - much closer to what we want.

There is definitely more room for optimization on that 1.94x speedup with:
- FSDP weight locality optimization
- Increasing gradient batch size
- ....

**Comment by pawalt** [comment]
@prbot approve @thundergolfer
